Bump to GHC 8.10.4

Anviking opened the pull request:
# Issue Number

ADP-453

# Overview

- [x] Bump to GHC-8.10.4
- [x] Use OnDeleteCascade in new persistent version, instead of our workaround in the fork

##### Todo:

- [x] Remove `allow-newer`
- [x] core:unit tests pass locally
- [x] Look closer at what why we needed the persistent fork
   - Seems we can now simply use OnDeleteCascade
- [x] Integration tests pass locally
- [x] Investigate cardano-wallet:unit failures
- [x] Bump haskell.nix
- [x] Figure out how to use weeder -> seems green at least
- [x] Get CI green
- [x] Move to snapshot in cardano-haskell repo
- [x] Merge that first

# Comments

[Hydra jobset](https://hydra.iohk.io/jobset/Cardano/cardano-wallet-pr-2533)

rvl committed:
scripts/gen-hie.sh: Update to use cabal plan.json

rvl committed:
Use nixpkgs pin from iohk-nix (nixos-unstable branch)

Anviking committed:
Bump to ghc-8.10.4

Anviking commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -110,7 +111,7 @@ TxMeta
     txMetadata              W.TxMetadata Maybe  sql=data
     txMetaSlotExpires       SlotNo Maybe        sql=slot_expires
-    txMetaFee               Natural Maybe       sql=fee
+    txMetaFee               Word Maybe       sql=fee
```
☝️

Anviking replied:
The `Natural` instance were removed with this explanation: https://hackage.haskell.org/package/persistent-2.11.0.2/docs/Database-Persist-Types.html#t:OverflowNatural

I think we might as well use Word instead of OverflowNatural, although maybe we should use `Word64` in case we're not on a 64 bit system?

rvl replied:
Yep, `Word64` is good, which should map to `SqlInt64`, same as `Natural` did previously.

Anviking replied:
Changed to Word64

Anviking committed:
Bump Haskell.nix to latest master

Anviking committed:
Enable unsafe migrations

There was a unit test failure where PersistUnsafeMigrationException was
thrown from a migration where we dropped a table.

Since the migration in question did create a
backup table, I guess enabling unsafe migrations should preserve the
same behaviour as before.

Here is the error:

  test/unit/Cardano/Pool/DB/SqliteSpec.hs:60:5:
  1) Cardano.Pool.DB.Sqlite, Migration Regressions, 'migrate' an existing database from v2019-12-16 by creating it from scratch again. But only once.
       uncaught exception: PersistUnsafeMigrationException

       Database migration: manual intervention required.
       The unsafe actions are prefixed by '***' below:

           CREATE TABLE "internal_state"("id" INTEGER PRIMARY KEY,"last_gc_metadata" VARCHAR NULL);
           CREATE TABLE "settings"("id" INTEGER PRIMARY KEY,"metadata_source" VARCHAR NOT NULL);
           CREATE TABLE "block_headers"("slot" INTEGER NOT NULL,"header_hash" VARCHAR NOT NULL,"parent_header_hash" VARCHAR NOT NULL,"block_height" INTEGER NOT NULL, PRIMARY KEY ("block_height"));
           CREATE TEMP TABLE "pool_owner_backup"("pool_id" VARCHAR NOT NULL,"slot" INTEGER NOT NULL,"slot_internal_index" INTEGER NOT NULL,"pool_owner" VARCHAR NOT NULL,"pool_owner_index" INTEGER NOT NULL, PRIMARY KEY ("pool_id","slot","slot_internal_index","pool_owner","pool_owner_index"));
           INSERT INTO "pool_owner_backup"("pool_id","pool_owner","pool_owner_index") SELECT "pool_id","pool_owner","pool_owner_index" FROM "pool_owner";
           DROP TABLE "pool_owner";
           CREATE TABLE "pool_owner"("pool_id" VARCHAR NOT NULL,"slot" INTEGER NOT NULL,"slot_internal_index" INTEGER NOT NULL,"pool_owner" VARCHAR NOT NULL,"pool_owner_index" INTEGER NOT NULL, PRIMARY KEY ("pool_id","slot","slot_internal_index","pool_owner","pool_owner_index"), CONSTRAINT "pool_ownerfk_registration_pool_id" FOREIGN KEY("pool_id","slot","slot_internal_index") REFERENCES "pool_registration"("pool_id","slot","slot_internal_index") ON DELETE CASCADE);
           INSERT INTO "pool_owner" SELECT "pool_id","slot","slot_internal_index","pool_owner","pool_owner_index" FROM "pool_owner_backup";
           DROP TABLE "pool_owner_backup";
           CREATE TEMP TABLE "pool_registration_backup"("pool_id" VARCHAR NOT NULL,"slot" INTEGER NOT NULL,"slot_internal_index" INTEGER NOT NULL,"margin_numerator" INTEGER NOT NULL,"margin_denominator" INTEGER NOT NULL,"cost" INTEGER NOT NULL,"pledge" INTEGER NOT NULL,"metadata_url" VARCHAR NULL,"metadata_hash" VARCHAR NULL, PRIMARY KEY ("pool_id","slot","slot_internal_index"));
           INSERT INTO "pool_registration_backup"("pool_id","cost") SELECT "pool_id","cost" FROM "pool_registration";
       *** DROP TABLE "pool_registration";
           CREATE TABLE "pool_registration"("pool_id" VARCHAR NOT NULL,"slot" INTEGER NOT NULL,"slot_internal_index" INTEGER NOT NULL,"margin_numerator" INTEGER NOT NULL,"margin_denominator" INTEGER NOT NULL,"cost" INTEGER NOT NULL,"pledge" INTEGER NOT NULL,"metadata_url" VARCHAR NULL,"metadata_hash" VARCHAR NULL, PRIMARY KEY ("pool_id","slot","slot_internal_index"));
           INSERT INTO "pool_registration" SELECT "pool_id","slot","slot_internal_index","margin_numerator","margin_denominator","cost","pledge","metadata_url","metadata_hash" FROM "pool_registration_backup";
           DROP TABLE "pool_registration_backup";
           CREATE TABLE "pool_delistment"("pool_id" VARCHAR NOT NULL, PRIMARY KEY ("pool_id"));
           CREATE TABLE "pool_retirement"("pool_id" VARCHAR NOT NULL,"slot" INTEGER NOT NULL,"slot_internal_index" INTEGER NOT NULL,"epoch" INTEGER NOT NULL, PRIMARY KEY ("pool_id","slot","slot_internal_index"));
           CREATE TABLE "pool_metadata"("metadata_hash" VARCHAR NOT NULL,"name" VARCHAR NOT NULL,"ticker" VARCHAR NOT NULL,"description" VARCHAR NULL,"homepage" VARCHAR NOT NULL, PRIMARY KEY ("metadata_hash"));
           CREATE TABLE "pool_metadata_fetch_attempts"("metadata_hash" VARCHAR NOT NULL,"metadata_url" VARCHAR NOT NULL,"retry_after" TIMESTAMP NOT NULL,"retry_count" INTEGER NOT NULL, PRIMARY KEY ("metadata_hash","metadata_url"));

  To rerun use: --match "/Cardano.Pool.DB.Sqlite/Migration Regressions/'migrate' an existing database from v2019-12-16 by creating it from scratch again. But only once./"

Anviking committed:
Leverage new OnDeleteCascade

OnDeleteCascade was introduced in the new version of persistent (2.11.0.0).

With it, we /shouldn't/ need the iohk fork anymore, and can replace the
! ON DELETE CASCADE.

Anviking commented on lib/core/test/unit/Cardano/Wallet/Api/Server/TlsSpec.hs:
```diff
@@ -2,6 +2,7 @@
 {-# LANGUAGE NamedFieldPuns #-}
 {-# LANGUAGE TemplateHaskell #-}
 
+{-# OPTIONS_GHC -Wno-missing-fields #-}
```
Don't think this should be here

rvl replied:
Yeah it seems like a helpful warning.

Anviking replied:
Think the HLS quick-fix actions added them (and the StartIsType ones)

(removed)

Anviking commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -42,15 +42,15 @@
 -- This module follows the article and example code (Version1.hs), pretty much
 -- exactly.
 
+{-# OPTIONS_GHC -Wno-missing-fields #-}
```
Don't think this should be here

Anviking commented on lib/core/src/Cardano/Wallet/Api/Client.hs:
```diff
@@ -330,9 +330,10 @@ addressClient =
-            , putRandomAddress  = \_ _ -> fail "feature unavailable."
-            , putRandomAddresses = \_ _ -> fail "feature unavailable."
+            , postRandomAddress = \_ _ -> error "feature unavailable."
+                -- TODO: Fix. Do we want a JSON error response?
```
I suspect we don't want `error` here

rvl replied:
Neither `error` nor `fail`.

`UnliftIO.Exception.throwString` is more explicit, yet still convenient to use in these circumstance.

Anviking commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -765,10 +765,14 @@ instance PersistField POSIXTime where
     fromPersistValue (PersistText time) =
         utcTimeToPOSIXSeconds <$>
-            parseTimeM True defaultTimeLocale
+            parseTimeM False defaultTimeLocale
```
Don't recall why I did this. 🤔  Should try reverting.

rvl replied:
Probably either value would work fine, but if in doubt, don't make parsing more strict.

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Script.hs:
```diff
@@ -158,16 +160,16 @@ constructAddressFromIx pTemplate dTemplate ix =
-    => KeyFingerprint "payment" k
-    -> Address
+    :: forall (n :: NetworkDiscriminant) (k :: Depth -> Type -> Type). Typeable n
+    => KeyFingerprint "payment" k -- ^
```
```suggestion
    => KeyFingerprint "payment" k
```

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -29,6 +29,7 @@
 -- a single account. In practice, one wants to manage a set of pools, one per
 -- account.
 
+{-# LANGUAGE StarIsType #-}
```
```suggestion
```

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite/TH.hs:
```diff
@@ -85,7 +84,7 @@ PrivateKey                             sql=private_key
 
     Primary privateKeyWalletId
-    Foreign Wallet fk_wallet_private_key privateKeyWalletId ! ON DELETE CASCADE
+    Foreign Wallet OnDeleteCascade fk_wallet_private_key privateKeyWalletId
```
Would you be able to make a fresh database and inspect the output of `.schema private_key`?
There should be an `ON DELETE CASCADE` in the wallet fk.

Anviking replied:
```sql
sqlite3 wallet-db/she.908bd12c214e1883b6ba1ca2cb5d7066a6a2cc30.sqlite
SQLite version 3.33.0 2020-08-14 13:23:32
Enter ".help" for usage hints.
sqlite>
sqlite> .schema private_key
CREATE TABLE IF NOT EXISTS "private_key"("wallet_id" VARCHAR NOT NULL,"root" BLOB NOT NULL,"hash" BLOB NOT NULL, PRIMARY KEY ("wallet_id"), CONSTRAINT "private_keyfk_wallet_private_key" FOREIGN KEY("wallet_id") REFERENCES "wallet"("wallet_id") ON DELETE CASCADE);
```

Anviking replied:
Seems to be the same as on an old DB:
```sql
sqlite3 wallet-db4/she.0e016aeb9828d1f61d5192b6e1bdea69f3d52017.sqlite
SQLite version 3.33.0 2020-08-14 13:23:32
Enter ".help" for usage hints.
sqlite> .schema private_key
CREATE TABLE IF NOT EXISTS "private_key"("wallet_id" VARCHAR NOT NULL,"root" BLOB NOT NULL,"hash" BLOB NOT NULL, PRIMARY KEY ("wallet_id"), CONSTRAINT "private_keyfk_wallet_private_key" FOREIGN KEY("wallet_id") REFERENCES "wallet"("wallet_id") ON DELETE CASCADE);
sqlite>
```

Anviking replied:
> We need to double check auto db migrations still work.

Starting the wallet on some old DB worked. And the unit tests failed when the new unsafe migrations weren't enabled.

That should be enough?

KtorZ replied:
If you're talking about these unit tests:

https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/test/unit/Cardano/Wallet/DB/SqliteSpec.hs#L262-L406

Then I'd say that your arguments are pretty convincing yes.

rvl commented on lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs:
```diff
@@ -765,10 +765,14 @@ instance PersistField POSIXTime where
 instance PersistFieldSql POSIXTime where
     sqlType _ = sqlType (Proxy @Text)
+
+instance MonadFail (Either Text) where
```
Not too keen on this ... where is it used?

Anviking replied:
For
```haskell
parseTimeM :: forall (m :: * -> *) t. (MonadFail m, ParseTime t) => Bool -> TimeLocale -> String -> String -> m t
```

But we can go with a 
```haskell
-- | Newtype to get a MonadFail instance for @Either Text@.
--
-- We need it to use @parseTimeM@.
newtype EitherText a = EitherText { getEitherText :: Either Text a }
    deriving (Functor, Applicative, Monad) via (Either Text)

instance MonadFail EitherText where
    fail = EitherText . Left . T.pack
```
instead.

rvl commented on stack.yaml:
```diff
@@ -17,53 +17,75 @@ packages:
-    - persistent-sqlite
-    - persistent-template
+
+  # TODO: We presumably need this fork still?
```
Nope!

Anviking replied:
Cleaned up

rvl reviewed (commented):
Looks pretty good so far, :tada:.
We need to double check auto db migrations still work.
I'll work on the haskell.nix bump.

rvl committed:
Bump iohk-nix to latest master

Anviking committed:
Fixup: Change fee DB type from Word to Word64

rvl committed:
Remove source-repository-package entries from cabal.project for now

The cabal.project can be regenerated later.

rvl committed:
Update hackage.nix and stackage.nix

rvl committed:
stack.yaml: Remove fixes which are no longer needed

rvl committed:
stack.yaml: iohk-monitoring back to rev from cardano-node-1.25.1

rvl committed:
nix: Use qualified imports of library functions

rvl committed:
Update nix-shell for Haskell.nix changes

Since we just bumped ghc and Haskell.nix, why not get the latest toys too?

haskell-language-server is updated from 0.6.0 to 1.0.0.0.

The materialization stuff is improved slightly as well.

rvl committed:
Quash another warning

rvl committed:
stack.yaml: update deps for new lts-17.5

rvl committed:
Run stack2cabal

rvl committed:
stylish-haskell fixes

rvl committed:
Fix for newer generic-lens

rvl committed:
Quash new warnings

rvl committed:
nix: Drop 19.03 and 20.09 nixpkgs pins

rvl committed:
nix: Changes for new Haskell.nix revision

Anviking committed:
Update CLISpec whitespace

Anviking committed:
Adjust estimateMaxInputsTests

They use arbitrary values generated with specific seeds. Something must
have changed in the arbitrary instances during the GHC bump, so we
should be able to simply change the expected values.

Anviking committed:
Try adding basic weeder.dhall

KtorZ commented on lib/core/src/Ouroboros/Network/Client/Wallet.hs:
```diff
@@ -222,7 +224,7 @@ chainSyncFollowTip toCardanoEra onTipUpdate =
 --
 -- See also 'send' for invoking commands.
-data ChainSyncCmd block (m :: * -> *)
+data ChainSyncCmd block (m :: Type -> Type)
```
Starting to wonder if enabling `{-# LANGUAGE StarIsType #-}` as a default extension wouldn't have made your life easier here :sweat_smile: ... I guess going over this change is now for the better.

KtorZ commented on lib/shelley/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -431,6 +431,9 @@ estimateMaxInputsTests cases = do
             it ("order of magnitude, nOuts = " <> o <> " => nInps = " <> i) $ do
+                -- NOTE: These tests broke in the GHC 8.6 -> 8.10 bump,
+                -- presumably due to some change in the arbitrary generation.
+                -- It would be better if they weren't so fragile.
```
We could perhaps "dump" the generated values in files used as input data for the test?

Anviking replied:
Probably - don't think it'd be worth the effort though. Would be 50 outputs in the source file.

But I wonder if this wouldn't be nice to some day have:
- `50 arbitrary outputs -> at least X1 inputs, never more than X2 inputs`
- `50 fixed-size outputs -> always Y inputs`

i.e. properties rather than goldens.

Anviking commented on cardano-1.25.1.yaml:
```diff
@@ -0,0 +1,157 @@
+- hedgehog-1.0.2
+- hedgehog-corpus-0.2.0
+- hedgehog-quickcheck-0.1.1
+- hspec-2.7.0
```
I wanted to use a more recent hspec for the `beforeAllWith` function when trying a thing.

I think there's no point in having this pinned to an old version.

Maybe we shouldn't drop non-test-related pinned packages though, if we want them to perfectly match what's used in cardano-node.

rvl replied:
Yes, we may as well use the hspec-2.7.8 version from the LTS snapshot.

If we change hedgehog versions, that may possibly break the build. We may as well leave it if it works.

Anviking committed:
Fixes for cabal.project

rvl commented on cabal.project:
```diff
@@ -130,3 +130,72 @@ source-repository-package
+  tests: False
+
+package io-sim-classes
+  tests: False
```
This block probably isn't required here.

rvl commented on stack.yaml:
```diff
@@ -1,9 +1,9 @@
 # NOTE: Remember to update the version matrix in README.md when
 # bumping the Cardano version.
-resolver: https://raw.githubusercontent.com/input-output-hk/cardano-haskell/1c257e85cadb503ced0fe0a9ba757ae449baaf0b/snapshots/cardano-1.25.1.yaml
+resolver: cardano-1.25.1.yaml
```
Is it time to add the snapshot to cardano-haskell?

Anviking replied:
Want to see if CI's green after un-pinning hspec, will move after that

rvl commented on cabal.project:
```diff
@@ -137,163 +131,71 @@ source-repository-package
-  tests: False
-  benchmarks: False
-package iohk-monitoring
+-- The following is needed because Nix is doing something crazy.
```
This block of package options probably isn't necessary here.

Anviking replied:
It was. It fixed the error from `nix-shell nix/cabal-shell.nix --run "cabal v2-configure -frelease --enable-tests --enable-benchmarks"`. The first section wasn't needed though... but seems fine to me to keep.

rvl replied:
It seems like cabal ignores the global `tests: False` flag, and then disabling tests on a selection of difficult packages is enough to make it work. This probjem is not likely to be related to nix - because we are running cabal outside of nix-build.

rvl committed:
nix: Fix windows build failure

Anviking committed:
Use the LTS hspec versions

Anviking committed:
Move snapshot to cardano-haskell repo

Anviking committed:
Move snapshot to cardano-haskell repo

rvl committed:
Update hackage.nix and stackage.nix

rvl committed:
Update niv boilerplate to latest version

rvl committed:
Regenerate nix

rvl committed:
Fix cabal.project

rvl committed:
Try to fix hydra eval error by bumping Haskell.nix

Commit:
fixup! Regenerate nix
